packages/store/addon/-private/ts-interfaces/ember-data-json-api.ts

```diff
@@ -6,6 +6,18 @@ import { Dict } from './utils';
 */
 
 export type Meta = Dict<JSONValue>;
+export type LinkObject = { href: string; meta?: Dict<JSONValue> };
+export type Link = string | LinkObject;
+export interface Links {
+  related?: Link;
+  self?: Link;
+}
+export interface PaginationLinks extends Links {
+  first?: Link | null;
```

Comment:
Just so I'm clear: this is doing optional-or-`null` for conformity with the JSON API spec, which allows these to be left off *or* explicitly set to `null`?

Reply:
correct. Typescript has many of these breakdowns when it comes to typing JSON structures. `undefined` is not actually a valid JSON value, but both `null` and `not-present` ARE

Reply:
(and this is the fundamental reason why EmberData ends up with `null | undefined` throughout the system instead of just one or the other :'( )